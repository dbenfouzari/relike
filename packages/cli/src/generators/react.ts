import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
