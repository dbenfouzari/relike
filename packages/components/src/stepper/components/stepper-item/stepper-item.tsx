import classnames from "classnames";
import type { CSSProperties, FC } from "react";
import { createUseStyles } from "react-jss";

import Color from "../../../color";
import Icon from "../../../icon";
import Icons, { IconData } from "../../../icons";
import classes from "./stepper-item.module.scss";

/** Defines StepperItem props */
export interface StepperItemProps {
  /**
   * Title of the step. Mandatory, it's the step title.
   */
  title: string;
  /**
   * A small step description.
   */
  description?: string;
  /**
   * Subtitle to allow showing additional data.
   */
  subTitle?: string;
}

/** Defines internal stepper item props. */
export interface InternalStepperItemProps extends StepperItemProps {
  /** Is the stepper item current ? */
  isCurrent: boolean;
  /** Is the stepper item past ? */
  isPast: boolean;
  /** Stepper item index in array */
  index: number;
  /**
   * Primary StepperItem color.
   *
   * It defines :
   * - past & current icon border color
   * - past line color
   * - current icon color
   */
  primaryColor: Color;
  /**
   * Secondary StepperItem color.
   *
   * It defines :
   * - past description color
   * - subtitle color (time left in Storybook example)
   */
  secondaryColor: Color;
  /**
   * Disabled StepperItem color.
   *
   * It defines :
   * - next item title color
   * - next item description color
   * - next item icon color
   */
  disabledColor: Color;
  /**
   * Disabled StepperItem color.
   *
   * It defines current title and description color.
   */
  titleColor: Color;
}

/** Defines props to generate stepper item styles. */
type StepperItemStylesProps = Pick<
  InternalStepperItemProps,
  "primaryColor" | "secondaryColor" | "disabledColor" | "titleColor"
>;

const useStyles = createUseStyles({
  /**
   * Generates styles based on props
   *
   * @param {StepperItemStylesProps} props The styles props.
   * @example
   * wrapper({
   *   disabledColor: Colors.grey,
   *   secondaryColor: Colors.red,
   *   primaryColor: Colors.blue,
   *   titleColor: Colors.blue
   * })
   * @returns {CSSProperties} CSS rules
   */
  wrapper: ({ disabledColor, secondaryColor, primaryColor, titleColor }: StepperItemStylesProps) =>
    ({
      "--stepper-main-color": primaryColor.toRGBA(),
      "--stepper-secondary-color": secondaryColor.toRGBA(),
      "--stepper-disabled-color": disabledColor.toRGBA(),
      "--stepper-title-color": titleColor.toRGBA(),
    } as CSSProperties),
});

/**
 * Stepper Item
 *
 * @example
 * <StepperItem title="Test" />
 * @returns {JSX.Element} JSX element.
 */
export const StepperItem: FC<StepperItemProps> = () => null;

/**
 * The internal StepperItem. It is only used by Stepper component to render something,
 * but should not be used directly.
 *
 * @param {InternalStepperItemProps} props The props.
 * @example
 * <InternalStepperItem
 *   key={index}
 *   isCurrent={index === current}
 *   isPast={index < current}
 *   index={index}
 *   disabledColor={disabledColor}
 *   secondaryColor={secondaryColor}
 *   primaryColor={primaryColor}
 *   titleColor={titleColor}
 *   {...child.props}
 * />
 * @returns {JSX.Element} JSX Element.
 */
export const InternalStepperItem: FC<InternalStepperItemProps> = ({
  title,
  description,
  subTitle,
  isCurrent,
  isPast,
  index,
  disabledColor,
  secondaryColor,
  primaryColor,
  titleColor,
}) => {
  const styles = useStyles({
    primaryColor,
    secondaryColor,
    disabledColor,
    titleColor,
  });

  const icon: IconData = isPast ? Icons.check : isCurrent ? Icons.check_box : Icons.pin;

  const iconComponent = isPast ? <Icon icon={icon} color={primaryColor} /> : index + 1;

  return (
    <div
      className={classnames(styles.wrapper, classes.item, {
        [classes.item__current]: isCurrent,
        [classes.item__past]: isPast,
      })}
    >
      <div className={classes.container}>
        <span className={classes.icon}>{iconComponent}</span>
        <div className={classes.content}>
          <span className={classes.title}>
            {title} <span className={classes.subtitle}>{subTitle}</span>
          </span>
          <span className={classes.description}>{description}</span>
        </div>
      </div>
    </div>
  );
};

export default StepperItem;
