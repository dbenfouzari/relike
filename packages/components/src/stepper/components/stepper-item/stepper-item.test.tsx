import { render, screen } from "@testing-library/react";

import Colors from "../../../colors";
import { InternalStepperItem, InternalStepperItemProps, StepperItem } from "./stepper-item";

/**
 * Set up InternalStepperItem tests
 *
 * @param   {Partial<InternalStepperItemProps>} props The props
 * @returns {void}                                    The result
 * @example
 * setup()
 */
const setup = (props: Partial<InternalStepperItemProps> = {}) => {
  const mergedProps: InternalStepperItemProps = {
    title: "Test",
    isCurrent: true,
    isPast: false,
    titleColor: Colors.blue,
    primaryColor: Colors.blue,
    secondaryColor: Colors.red,
    disabledColor: Colors.grey,
    index: 1,
    ...props,
  };

  render(<InternalStepperItem {...mergedProps} />);
};

describe("Components/StepperItem", () => {
  it("StepperItem should render null", () => {
    const { container } = render(<StepperItem title="Zbra" />);

    // Disable ESLint because I want to test that it renders nothing.
    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toBe(null);
  });

  it("should render successfully", () => {
    setup();

    expect(screen.getByRole("listitem")).toMatchSnapshot();
  });

  it("should render successfully when is past", () => {
    setup({ isPast: true });

    expect(screen.getByRole("listitem")).toMatchSnapshot();
  });
});
