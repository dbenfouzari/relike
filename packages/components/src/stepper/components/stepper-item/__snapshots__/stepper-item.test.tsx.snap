// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/StepperItem should render successfully 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d0-0-2-2 item item__current"
  role="listitem"
>
  <div
    class="container"
  >
    <span
      class="icon"
    >
      2
    </span>
    <div
      class="content"
    >
      <span
        class="title"
      >
        Test
         
        <span
          class="subtitle"
        />
      </span>
      <span
        class="description"
      />
    </div>
  </div>
</div>
`;

exports[`Components/StepperItem should render successfully when is past 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d1-0-2-3 item item__current item__past"
  role="listitem"
>
  <div
    class="container"
  >
    <span
      class="icon"
    >
      <span
        class="jsx-1118668747 icon material-icons"
        role="img"
      >
        check
      </span>
    </span>
    <div
      class="content"
    >
      <span
        class="title"
      >
        Test
         
        <span
          class="subtitle"
        />
      </span>
      <span
        class="description"
      />
    </div>
  </div>
</div>
`;
